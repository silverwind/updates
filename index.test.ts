import spawn from "nano-spawn";
import restana from "restana";
import {join, dirname} from "node:path";
import {readFileSync, mkdtempSync} from "node:fs";
import {writeFile, readFile, rm} from "node:fs/promises";
import {fileURLToPath} from "node:url";
import {tmpdir} from "node:os";
import {env} from "node:process";
import type {Server} from "node:http";
import type {Service, Protocol} from "restana";

const testFile = fileURLToPath(new URL("fixtures/npm-test/package.json", import.meta.url));
const emptyFile = fileURLToPath(new URL("fixtures/npm-empty/package.json", import.meta.url));
const poetryFile = fileURLToPath(new URL("fixtures/poetry/pyproject.toml", import.meta.url));
const dualFile = fileURLToPath(new URL("fixtures/dual", import.meta.url));

const testPkg = JSON.parse(readFileSync(testFile, "utf8"));
const testDir = mkdtempSync(join(tmpdir(), "updates-"));
const script = fileURLToPath(new URL("dist/index.js", import.meta.url));

const dependencyTypes = [
  "dependencies",
  "devDependencies",
  "peerDependencies",
  "optionalDependencies",
  "resolutions",
];

const testPackages: Set<string> = new Set();
for (const dependencyType of dependencyTypes) {
  for (const name of Object.keys(testPkg[dependencyType] || [])) {
    testPackages.add(name);
  }
}

const pyTestPackages = new Set(["djlint", "PyYAML"]);

function makeUrl(server: Server) {
  const {port}: any = server.address();
  return Object.assign(new URL("http://localhost"), {port}).toString();
}

function defaultRoute(req: any, res: any) {
  console.error(`default handler hit for ${req.url}`);
  res.send(404);
}

function resolutionsBasePackage(name: string) {
  const packages = name.match(/(@[^/]+\/)?([^/]+)/g) || [];
  return packages[packages.length - 1];
}

let npmServer: Service<Protocol.HTTP> | Server;
let githubServer: Service<Protocol.HTTP> | Server;
let pypiServer: Service<Protocol.HTTP> | Server;

let githubUrl: string;
let pypiUrl: string;
let npmUrl: string;

beforeAll(async () => {
  let commits, tags;

  [npmServer, githubServer, pypiServer, commits, tags] = await Promise.all([
    restana({defaultRoute}),
    restana({defaultRoute}),
    restana({defaultRoute}),
    readFile(fileURLToPath(new URL("fixtures/github/updates-commits.json", import.meta.url))),
    readFile(fileURLToPath(new URL("fixtures/github/updates-tags.json", import.meta.url))),
  ]);

  for (const pkgName of testPackages) {
    const name = testPkg.resolutions[pkgName] ? resolutionsBasePackage(pkgName) : pkgName;
    const urlName = name.replace(/\//g, "%2f");
    // can not use file URLs because node stupidely throws on "%2f" in paths.
    const path = join(dirname(fileURLToPath(import.meta.url)), `fixtures/npm/${urlName}.json`);
    npmServer.get(`/${urlName}`, async (_, res) => res.send(await readFile(path)));
  }

  for (const pkgName of pyTestPackages) {
    const path = join(dirname(fileURLToPath(import.meta.url)), `fixtures/pypi/${pkgName}.json`);
    pypiServer.get(`/pypi/${pkgName}/json`, async (_, res) => res.send(await readFile(path)));
  }

  githubServer.get("/repos/silverwind/updates/commits", (_, res) => res.send(commits));
  githubServer.get("/repos/silverwind/updates/git/refs/tags", (_, res) => res.send(tags));

  [githubServer, pypiServer, npmServer] = await Promise.all([
    githubServer.start(0),
    pypiServer.start(0),
    npmServer.start(0),
  ]);

  githubUrl = makeUrl(githubServer);
  npmUrl = makeUrl(npmServer);
  pypiUrl = makeUrl(pypiServer);

  await writeFile(join(testDir, ".npmrc"), `registry=${npmUrl}`); // Fake registry
  await writeFile(join(testDir, "package.json"), JSON.stringify(testPkg, null, 2)); // Copy fixture
});

afterAll(async () => {
  await Promise.all([
    rm(testDir, {recursive: true}),
    npmServer?.close(),
    githubServer?.close(),
  ]);
});

function makeTest(args: string) {
  return async () => {
    const argsArr = [
      ...args.split(/\s+/), "-c",
      "--githubapi", githubUrl,
      "--pypiapi", pypiUrl,
    ];
    const {stdout} = await spawn(process.execPath, [script, ...argsArr], {cwd: testDir});
    const {results} = JSON.parse(stdout);

    // Parse results, with custom validation for the dynamic "age" property
    for (const mode of Object.keys(results || {})) {
      for (const dependencyType of [
        ...dependencyTypes,
        "tool.poetry.dependencies",
        "tool.poetry.dev-dependencies",
        "tool.poetry.test-dependencies",
        "tool.poetry.group.dev.dependencies",
        "tool.poetry.group.test.dependencies",
      ]) {
        for (const name of Object.keys(results?.[mode]?.[dependencyType] || {})) {
          delete results[mode][dependencyType][name].age;
        }
      }
    }

    expect(results).toMatchSnapshot();
  };
}

test("simple", async () => {
  const {stdout, stderr} = await spawn(process.execPath, [
    script,
    "-C",
    "--githubapi", githubUrl,
    "--pypiapi", pypiUrl,
    "--registry", npmUrl,
    "-f", testFile,
  ]);
  expect(stderr).toEqual("");
  expect(stdout).toContain("prismjs");
  expect(stdout).toContain("https://github.com/silverwind/updates");
});

test("empty", async () => {
  const {stdout, stderr} = await spawn(process.execPath, [
    script,
    "-C",
    "--githubapi", githubUrl,
    "--pypiapi", pypiUrl,
    "-f", emptyFile,
  ]);
  expect(stderr).toEqual("");
  expect(stdout).toContain("No dependencies");
});

if (env.CI && !env.BUN) {
  test("global", async () => {
    await spawn("npm", ["i", "-g", "."]);
    const {stdout, stderr} = await spawn("updates", [
      "-C",
      "--githubapi", githubUrl,
      "--pypiapi", pypiUrl,
      "-f", testFile,
    ]);
    expect(stderr).toEqual("");
    expect(stdout).toContain("prismjs");
    expect(stdout).toContain("https://github.com/silverwind/updates");
  });
}

test("latest", makeTest("-j"));
test("greatest", makeTest("-j -g"));
test("prerelease", makeTest("-j -g -p"));
test("release", makeTest("-j -R"));
test("patch", makeTest("-j -P"));
test("include", makeTest("-j -i noty"));
test("include 2", makeTest("-j -i noty -i noty,noty"));
test("include 3", makeTest("-j -i /^noty/"));
test("exclude", makeTest("-j -e gulp-sourcemaps,prismjs,svgstore,html-webpack-plugin,noty,jpeg-buffer-orientation,styled-components,@babel/preset-env,versions/updates,react"));
test("exclude 2", makeTest("-j -e gulp-sourcemaps -i /react/"));
test("exclude 3", makeTest("-j -i gulp*"));
test("exclude 4", makeTest("-j -i /^gulp/ -P gulp*"));
test("pypi", makeTest(`-j -f ${poetryFile}`));
test("dual", makeTest(`-j -f ${dualFile}`));
test("dual 2", makeTest(`-j -f ${dualFile} -i noty`));
