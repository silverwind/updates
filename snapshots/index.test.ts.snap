// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dual 1`] = `
{
  "npm": {
    "dependencies": {
      "@babel/preset-env": {
        "info": "https://github.com/babel/babel/tree/HEAD/packages/babel-preset-env",
        "new": "7.11.5",
        "old": "7.0.0",
      },
      "gulp-sourcemaps": {
        "info": "https://github.com/gulp-sourcemaps/gulp-sourcemaps",
        "new": "2.6.5",
        "old": "2.0.0",
      },
      "html-webpack-plugin": {
        "info": "https://github.com/jantimon/html-webpack-plugin",
        "new": "4.0.0-beta.11",
        "old": "4.0.0-alpha.2",
      },
      "jpeg-buffer-orientation": {
        "info": "https://github.com/fisker/jpeg-buffer-orientation",
        "new": "2.0.3",
        "old": "0.0.0",
      },
      "noty": {
        "info": "https://github.com/needim/noty",
        "new": "3.2.0-beta",
        "old": "3.1.0",
      },
      "prismjs": {
        "info": "https://github.com/LeaVerou/prism",
        "new": "1.17.1",
        "old": "1.0.0",
      },
      "react": {
        "info": "https://github.com/facebook/react/tree/HEAD/packages/react",
        "new": "18.2.0",
        "old": "18.0",
      },
      "styled-components": {
        "info": "https://github.com/styled-components/styled-components",
        "new": "5.0.0-rc.2",
        "old": "2.5.0-1",
      },
      "svgstore": {
        "info": "https://github.com/svgstore/svgstore",
        "new": "^3.0.0-2",
        "old": "^3.0.0",
      },
      "updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "537ccb7",
        "old": "6941e05",
      },
    },
    "peerDependencies": {
      "@babel/preset-env": {
        "info": "https://github.com/babel/babel/tree/HEAD/packages/babel-preset-env",
        "new": "~7.11.5",
        "old": "~6.0.0",
      },
    },
    "resolutions": {
      "versions/updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "^10.0.0",
        "old": "^1.0.0",
      },
    },
  },
  "pypi": {
    "dependencies": {
      "updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "537ccb7",
        "old": "6941e05",
      },
    },
    "tool.poetry.group.dev.dependencies": {
      "djlint": {
        "info": "https://github.com/Riverside-Healthcare/djlint",
        "new": "1.31.0",
        "old": "1.30.0",
      },
    },
  },
}
`;

exports[`dual 2 1`] = `
{
  "npm": {
    "dependencies": {
      "noty": {
        "info": "https://github.com/needim/noty",
        "new": "3.2.0-beta",
        "old": "3.1.0",
      },
    },
  },
}
`;

exports[`exclude 1`] = `
{
  "npm": {
    "dependencies": {
      "updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "537ccb7",
        "old": "6941e05",
      },
    },
    "peerDependencies": {
      "typescript": {
        "info": "https://github.com/Microsoft/TypeScript",
        "new": "^5",
        "old": "^4",
      },
    },
  },
}
`;

exports[`exclude 2 1`] = `
{
  "npm": {
    "dependencies": {
      "react": {
        "info": "https://github.com/facebook/react/tree/HEAD/packages/react",
        "new": "18.2.0",
        "old": "18.0",
      },
    },
  },
}
`;

exports[`exclude 3 1`] = `
{
  "npm": {
    "dependencies": {
      "gulp-sourcemaps": {
        "info": "https://github.com/gulp-sourcemaps/gulp-sourcemaps",
        "new": "2.6.5",
        "old": "2.0.0",
      },
    },
  },
}
`;

exports[`exclude 4 1`] = `
{
  "npm": {
    "dependencies": {
      "gulp-sourcemaps": {
        "info": "https://github.com/floridoo/gulp-sourcemaps",
        "new": "2.0.1",
        "old": "2.0.0",
      },
    },
  },
}
`;

exports[`greatest 1`] = `
{
  "npm": {
    "dependencies": {
      "@babel/preset-env": {
        "info": "https://github.com/babel/babel/tree/HEAD/packages/babel-preset-env",
        "new": "7.11.5",
        "old": "7.0.0",
      },
      "gulp-sourcemaps": {
        "info": "https://github.com/gulp-sourcemaps/gulp-sourcemaps",
        "new": "2.6.5",
        "old": "2.0.0",
      },
      "html-webpack-plugin": {
        "info": "https://github.com/jantimon/html-webpack-plugin",
        "new": "4.0.0-beta.11",
        "old": "4.0.0-alpha.2",
      },
      "jpeg-buffer-orientation": {
        "info": "https://github.com/fisker/jpeg-buffer-orientation",
        "new": "2.0.3",
        "old": "0.0.0",
      },
      "noty": {
        "info": "https://github.com/needim/noty",
        "new": "3.1.4",
        "old": "3.1.0",
      },
      "prismjs": {
        "info": "https://github.com/LeaVerou/prism",
        "new": "1.17.1",
        "old": "1.0.0",
      },
      "react": {
        "info": "https://github.com/facebook/react/tree/HEAD/packages/react",
        "new": "18.2.0",
        "old": "18.0",
      },
      "styled-components": {
        "info": "https://github.com/styled-components/styled-components",
        "new": "5.0.0-rc.2",
        "old": "2.5.0-1",
      },
      "updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "537ccb7",
        "old": "6941e05",
      },
    },
    "peerDependencies": {
      "@babel/preset-env": {
        "info": "https://github.com/babel/babel/tree/HEAD/packages/babel-preset-env",
        "new": "~7.11.5",
        "old": "~6.0.0",
      },
      "typescript": {
        "info": "https://github.com/Microsoft/TypeScript",
        "new": "^5",
        "old": "^4",
      },
    },
    "resolutions": {
      "versions/updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "^10.0.0",
        "old": "^1.0.0",
      },
    },
  },
}
`;

exports[`include 1`] = `
{
  "npm": {
    "dependencies": {
      "noty": {
        "info": "https://github.com/needim/noty",
        "new": "3.2.0-beta",
        "old": "3.1.0",
      },
    },
  },
}
`;

exports[`include 2 1`] = `
{
  "npm": {
    "dependencies": {
      "noty": {
        "info": "https://github.com/needim/noty",
        "new": "3.2.0-beta",
        "old": "3.1.0",
      },
    },
  },
}
`;

exports[`include 3 1`] = `
{
  "npm": {
    "dependencies": {
      "noty": {
        "info": "https://github.com/needim/noty",
        "new": "3.2.0-beta",
        "old": "3.1.0",
      },
    },
  },
}
`;

exports[`latest 1`] = `
{
  "npm": {
    "dependencies": {
      "@babel/preset-env": {
        "info": "https://github.com/babel/babel/tree/HEAD/packages/babel-preset-env",
        "new": "7.11.5",
        "old": "7.0.0",
      },
      "gulp-sourcemaps": {
        "info": "https://github.com/gulp-sourcemaps/gulp-sourcemaps",
        "new": "2.6.5",
        "old": "2.0.0",
      },
      "html-webpack-plugin": {
        "info": "https://github.com/jantimon/html-webpack-plugin",
        "new": "4.0.0-beta.11",
        "old": "4.0.0-alpha.2",
      },
      "jpeg-buffer-orientation": {
        "info": "https://github.com/fisker/jpeg-buffer-orientation",
        "new": "2.0.3",
        "old": "0.0.0",
      },
      "noty": {
        "info": "https://github.com/needim/noty",
        "new": "3.2.0-beta",
        "old": "3.1.0",
      },
      "prismjs": {
        "info": "https://github.com/LeaVerou/prism",
        "new": "1.17.1",
        "old": "1.0.0",
      },
      "react": {
        "info": "https://github.com/facebook/react/tree/HEAD/packages/react",
        "new": "18.2.0",
        "old": "18.0",
      },
      "styled-components": {
        "info": "https://github.com/styled-components/styled-components",
        "new": "5.0.0-rc.2",
        "old": "2.5.0-1",
      },
      "svgstore": {
        "info": "https://github.com/svgstore/svgstore",
        "new": "^3.0.0-2",
        "old": "^3.0.0",
      },
      "updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "537ccb7",
        "old": "6941e05",
      },
    },
    "peerDependencies": {
      "@babel/preset-env": {
        "info": "https://github.com/babel/babel/tree/HEAD/packages/babel-preset-env",
        "new": "~7.11.5",
        "old": "~6.0.0",
      },
      "typescript": {
        "info": "https://github.com/Microsoft/TypeScript",
        "new": "^5",
        "old": "^4",
      },
    },
    "resolutions": {
      "versions/updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "^10.0.0",
        "old": "^1.0.0",
      },
    },
  },
}
`;

exports[`patch 1`] = `
{
  "npm": {
    "dependencies": {
      "gulp-sourcemaps": {
        "info": "https://github.com/floridoo/gulp-sourcemaps",
        "new": "2.0.1",
        "old": "2.0.0",
      },
      "html-webpack-plugin": {
        "info": "https://github.com/jantimon/html-webpack-plugin",
        "new": "4.0.0-beta.11",
        "old": "4.0.0-alpha.2",
      },
      "noty": {
        "info": "https://github.com/needim/noty",
        "new": "3.1.4",
        "old": "3.1.0",
      },
      "react": {
        "info": "https://github.com/facebook/react/tree/HEAD/packages/react",
        "new": "18.0.0",
        "old": "18.0",
      },
      "svgstore": {
        "info": "https://github.com/svgstore/svgstore",
        "new": "^3.0.0-2",
        "old": "^3.0.0",
      },
      "updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "537ccb7",
        "old": "6941e05",
      },
    },
    "resolutions": {
      "versions/updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "^1.0.6",
        "old": "^1.0.0",
      },
    },
  },
}
`;

exports[`poetry 1`] = `
{
  "pypi": {
    "tool.poetry.group.dev.dependencies": {
      "PyYAML": {
        "info": "https://github.com/yaml/pyyaml",
        "new": "6.0",
        "old": "1.0",
      },
      "djlint": {
        "info": "https://github.com/Riverside-Healthcare/djlint",
        "new": "1.31.0",
        "old": "1.30.0",
      },
    },
  },
}
`;

exports[`prerelease 1`] = `
{
  "npm": {
    "dependencies": {
      "@babel/preset-env": {
        "info": "https://github.com/babel/babel/tree/HEAD/packages/babel-preset-env",
        "new": "7.11.5",
        "old": "7.0.0",
      },
      "gulp-sourcemaps": {
        "info": "https://github.com/gulp-sourcemaps/gulp-sourcemaps",
        "new": "2.6.5",
        "old": "2.0.0",
      },
      "html-webpack-plugin": {
        "info": "https://github.com/jantimon/html-webpack-plugin",
        "new": "4.0.0-beta.11",
        "old": "4.0.0-alpha.2",
      },
      "jpeg-buffer-orientation": {
        "info": "https://github.com/fisker/jpeg-buffer-orientation",
        "new": "2.0.3",
        "old": "0.0.0",
      },
      "noty": {
        "info": "https://github.com/needim/noty",
        "new": "3.2.0-beta",
        "old": "3.1.0",
      },
      "prismjs": {
        "info": "https://github.com/LeaVerou/prism",
        "new": "1.17.1",
        "old": "1.0.0",
      },
      "react": {
        "info": "https://github.com/facebook/react/tree/HEAD/packages/react",
        "new": "18.3.0-next-d1e35c703-20221110",
        "old": "18.0",
      },
      "styled-components": {
        "info": "https://github.com/styled-components/styled-components",
        "new": "5.0.0-rc.2",
        "old": "2.5.0-1",
      },
      "svgstore": {
        "info": "https://github.com/svgstore/svgstore",
        "new": "^3.0.0-2",
        "old": "^3.0.0",
      },
      "updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "537ccb7",
        "old": "6941e05",
      },
    },
    "peerDependencies": {
      "@babel/preset-env": {
        "info": "https://github.com/babel/babel/tree/HEAD/packages/babel-preset-env",
        "new": "~7.11.5",
        "old": "~6.0.0",
      },
      "typescript": {
        "info": "https://github.com/Microsoft/TypeScript",
        "new": "^5",
        "old": "^4",
      },
    },
    "resolutions": {
      "versions/updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "^10.0.0",
        "old": "^1.0.0",
      },
    },
  },
}
`;

exports[`release 1`] = `
{
  "npm": {
    "dependencies": {
      "@babel/preset-env": {
        "info": "https://github.com/babel/babel/tree/HEAD/packages/babel-preset-env",
        "new": "7.11.5",
        "old": "7.0.0",
      },
      "gulp-sourcemaps": {
        "info": "https://github.com/gulp-sourcemaps/gulp-sourcemaps",
        "new": "2.6.5",
        "old": "2.0.0",
      },
      "html-webpack-plugin": {
        "info": "https://github.com/jantimon/html-webpack-plugin",
        "new": "3.2.0",
        "old": "4.0.0-alpha.2",
      },
      "jpeg-buffer-orientation": {
        "info": "https://github.com/fisker/jpeg-buffer-orientation",
        "new": "2.0.3",
        "old": "0.0.0",
      },
      "noty": {
        "info": "https://github.com/needim/noty",
        "new": "3.1.4",
        "old": "3.1.0",
      },
      "prismjs": {
        "info": "https://github.com/LeaVerou/prism",
        "new": "1.17.1",
        "old": "1.0.0",
      },
      "react": {
        "info": "https://github.com/facebook/react/tree/HEAD/packages/react",
        "new": "18.2.0",
        "old": "18.0",
      },
      "styled-components": {
        "info": "https://github.com/styled-components/styled-components",
        "new": "4.4.1",
        "old": "2.5.0-1",
      },
      "svgstore": {
        "info": "https://github.com/svgstore/svgstore",
        "new": "^2.0.3",
        "old": "^3.0.0",
      },
      "updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "537ccb7",
        "old": "6941e05",
      },
    },
    "peerDependencies": {
      "@babel/preset-env": {
        "info": "https://github.com/babel/babel/tree/HEAD/packages/babel-preset-env",
        "new": "~7.11.5",
        "old": "~6.0.0",
      },
      "typescript": {
        "info": "https://github.com/Microsoft/TypeScript",
        "new": "^5",
        "old": "^4",
      },
    },
    "resolutions": {
      "versions/updates": {
        "info": "https://github.com/silverwind/updates",
        "new": "^10.0.0",
        "old": "^1.0.0",
      },
    },
  },
}
`;

exports[`uv 1`] = `
{
  "pypi": {
    "dependency-groups.dev": {
      "PyYAML": {
        "age": "3 years",
        "info": "https://github.com/yaml/pyyaml",
        "new": "6.0",
        "old": "1.0",
      },
    },
    "project.dependencies": {
      "djlint": {
        "age": "2 years",
        "info": "https://github.com/Riverside-Healthcare/djlint",
        "new": "1.31.0",
        "old": "1.30.0",
      },
    },
  },
}
`;
